applications/StructuralMechanicsApplication/custom_response_functions/response_utilities/differentiation_utility.cpp

```diff
@@ -0,0 +1,258 @@
+
+            // perturb the design variable
+            const double current_property_value = rElement.GetProperties()[rDesignVariable];
+            p_local_property->SetValue(rDesignVariable, (current_property_value + rPertubationSize));
+
+            // Compute RHS after perturbation
+            rElement.CalculateRightHandSide(RHS_perturbed, rCurrentProcessInfo);
+
+            // Compute derivative of RHS w.r.t. design variable with finite differences
+            for(IndexType i = 0; i < RHS_perturbed.size(); ++i)
+                rOutput(0, i) = (RHS_perturbed[i] - RHS_unperturbed[i]) / rPertubationSize;
+
+            // Give element original properties back
+            rElement.SetProperties(p_global_properties);
+
+            //call one last time to make sure everything is as it was before TODO improve this..
+            rElement.CalculateRightHandSide(RHS_perturbed, rCurrentProcessInfo);
+        }
+        else
+            if ( (rOutput.size1() != 0) || (rOutput.size2() != 0) )
+                rOutput.resize(0,0,false);
+
+        KRATOS_CATCH("");
+    }
+
+    void DifferentiationUtility::CalculateRigthHandSideDerivative(Element& rElement,
+                                                const array_1d_component_type& rDesignVariable,
+                                                Node<3>& rNode,
+                                                const double& rPertubationSize,
+                                                Vector& rOutput,
+                                                ProcessInfo& rCurrentProcessInfo)
+    {
+        KRATOS_TRY;
+
+        #ifdef _OPENMP
+            KRATOS_ERROR_IF(omp_get_thread_num() > 0) <<
+                "DifferentiationUtility::CalculateRigthHandSideDerivative " <<
+                "is not thread safe for shape derivatives!" << omp_get_thread_num();
+        #endif
+
+        if( rDesignVariable == SHAPE_X || rDesignVariable == SHAPE_Y || rDesignVariable == SHAPE_Z )
+        {
+           const IndexType coord_dir = DifferentiationUtility::GetCoordinateDirection(rDesignVariable);
+
+            // define working variables
+            Vector RHS_unperturbed;
+            Vector RHS_perturbed;
+
+            // compute RHS before perturbion
+            rElement.CalculateRightHandSide(RHS_unperturbed, rCurrentProcessInfo);
+
+            if ( rOutput.size() != RHS_unperturbed.size() )
+                rOutput.resize(RHS_unperturbed.size(), false);
+
+            // perturb the design variable
+            rNode.GetInitialPosition()[coord_dir] += rPertubationSize;
+            rNode.Coordinates()[coord_dir] += rPertubationSize;
+
+            // compute LHS after perturbation
+            rElement.CalculateRightHandSide(RHS_perturbed, rCurrentProcessInfo);
+
+            //compute derivative of RHS w.r.t. design variable with finite differences
+            noalias(rOutput) = (RHS_perturbed - RHS_unperturbed) / rPertubationSize;
```

Comment:
we should also talk abt the migration to AMatrix, this also affects you in a mayor way

Reply:
Should we do this already within this PR? Is this even possible to migrate the AMatrix only locally in this class or is it necessary to modify the hole adjoint implementation. For the second case I would suggest to keep this PR as it is and to migrate the AMatrix in an upcoming PR.

Reply:
yes the migration should definitely be in a separate PR, we should speak abt it beforehand in person